etk-asm/src/ops/expression.rs

```diff
@@ -0,0 +1,420 @@
+
+    #[snafu(display("unknown macro `{}`", name))]
+    #[non_exhaustive]
+    UnknownMacro { name: String, backtrace: Backtrace },
+
+    #[snafu(display("undefined macro variable `{}`", name))]
+    #[non_exhaustive]
+    UndefinedVariable { name: String, backtrace: Backtrace },
+}
+
+type LabelsMap = HashMap<String, Option<u32>>;
+type VariablesMap = HashMap<String, Expression>;
+type MacrosMap = HashMap<String, MacroDefinition>;
+
+/// Evaluation context for `Expression`.
+#[derive(Clone, Copy, Debug)]
+pub struct Context<'a> {
+    labels: Option<&'a LabelsMap>,
+    macros: Option<&'a MacrosMap>,
+    variables: Option<&'a VariablesMap>,
+}
+
+impl<'a> Context<'a> {
+    /// Instantiates an empty `Context`.
+    pub fn empty() -> Self {
+        Context {
+            labels: None,
+            macros: None,
+            variables: None,
+        }
+    }
+
+    /// Looks up a label in the current context.
+    pub fn get_label(&self, key: &str) -> Option<&Option<u32>> {
+        match self.labels {
+            Some(labels) => labels.get(key),
+            None => None,
+        }
+    }
+
+    /// Looks up a macro in the current context.
+    pub fn get_macro(&self, key: &str) -> Option<&MacroDefinition> {
+        match self.macros {
+            Some(macros) => macros.get(key),
+            None => None,
+        }
+    }
+
+    /// Looks up a variable in the current context.
+    pub fn get_variable(&self, key: &str) -> Option<&Expression> {
+        match self.variables {
+            Some(variables) => variables.get(key),
+            None => None,
+        }
+    }
+}
+
+impl<'a> From<&'a LabelsMap> for Context<'a> {
+    fn from(labels: &'a LabelsMap) -> Self {
+        Self {
+            labels: Some(labels),
+            macros: None,
+            variables: None,
+        }
+    }
+}
+
+impl<'a> From<(&'a LabelsMap, &'a MacrosMap)> for Context<'a> {
+    fn from(x: (&'a LabelsMap, &'a MacrosMap)) -> Self {
+        Self {
+            labels: Some(x.0),
+            macros: Some(x.1),
+            variables: None,
+        }
+    }
+}
+
+impl<'a> From<(&'a LabelsMap, &'a MacrosMap, &'a VariablesMap)> for Context<'a> {
+    fn from(x: (&'a LabelsMap, &'a MacrosMap, &'a VariablesMap)) -> Self {
+        Self {
+            labels: Some(x.0),
+            macros: Some(x.1),
+            variables: Some(x.2),
+        }
+    }
+}
+
+/// An mathematical expression.
+#[derive(Clone, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub enum Expression {
+    /// A mathematical expression.
+    Expression(Box<Self>),
+
+    /// An expression macro invocation.
+    Macro(ExpressionMacroInvocation),
+
+    /// A terminal value.
+    Terminal(Terminal),
+
+    /// An addition operation.
+    Plus(Box<Self>, Box<Self>),
+
+    /// A subtraction operation.
+    Minus(Box<Self>, Box<Self>),
+
+    /// A multiplication operation.
+    Times(Box<Self>, Box<Self>),
+
+    /// A division operation.
+    Divide(Box<Self>, Box<Self>),
+}
```

Comment:
Leaving myself a reminder to maybe restructure this into a tree the same way [`Expr`](https://github.com/quilt/etk/blob/b2045abe61a1a8b1428f75af371c3f89b4959fa2/etk-dasm/src/sym.rs#L28) works.